Add/stats tracking pixel to pkg

darssen opened the pull request:
Spin off from: https://github.com/Automattic/jetpack/pull/26348
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Tracking pixel functionality is now handled by `Automattic\Jetpack\Stats\Tracking_Pixel`

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p9dueE-5Rd-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Look at the code and unit tests

darssen committed:
Stats package: Add tracking pixel class

darssen committed:
changelog

darssen commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,166 @@
+	 */
+	public static function add_to_footer() {
+		$data = self::build_view_data();
+		// TODO Is this really needed here?
```
We are still unsure if this is necessary since AMP implementation seems to not be using this code.

darssen replied:
We are still unsure if this is necessary since AMP implementation seems to not be using this code.

darssen commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,166 @@
+	 * @access public
+	 * @return void
+	 */
+	public static function add_to_footer() {
```
What would be the best way to test this functionality in this class?

brbrr replied:
As an option, we could extract composing part of the `render_*` functions into a `get_footer_*`, and then we would be able to test that expected footer to be generated. 

Additionally, we can have an arbitrary `get_footer` that would encapsulate AMP-related logic, this way, this function could look like this:

```
$data = self::build_view_data();
$footer = self::get_footer($data);
print $footer; // or render($footer);
```

Does it make sense?

darssen reviewed (commented):
We want to test the functionalities of both `render footer` and `render_amp_footer` with the `add_to_footer` static function since the former should be private methods. We are just keeping them public for backward compatibility.

Not sure what would be the best approach for testing that the HTML code has been inserted.

fgiannar commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,166 @@
+	 * @param array $kvs Array of options about the site and page you're on.
+	 * @return string
+	 */
+	public static function stats_array( $kvs ) {
```
We could rename this to `stats_array_to_string`

fgiannar commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,166 @@
+	 *
+	 * @var string $array_name The 'stats' array name
+	 */
+	const ARRAY_NAME = 'stats_array';
```
We can make it a bit more obvious what this is about, for example rename it to `STATS_ARRAY_TO_STRING_FILTER` and update its description accordingly :)

fgiannar commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,166 @@
+	 */
+	public static function stats_array( $kvs ) {
+
+		$kvs   = (array) apply_filters( self::ARRAY_NAME, $kvs );
```
It would be nice to add a docblock here about this filter, see [here](https://github.com/Automattic/jetpack/blob/d38405b569aad92c4164b61acb57b884dc11b2c0/projects/plugins/jetpack/modules/stats.php#L460):
```
/**
         * Filters the options added to the JavaScript Stats tracking code.
         *
         * @since-jetpack 1.1.0
         *
         * @param array $kvs Array of options about the site and page you're on.
         */
  ```

fgiannar commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,166 @@
+		$v        = 'ext';
+		$blog_url = wp_parse_url( site_url() );
+		$srv      = $blog_url['host'];
+		$j        = sprintf( '%s:%s', Constants::get_constant( 'JETPACK__API_VERSION' ), Constants::get_constant( 'JETPACK__VERSION' ) );
```
I'm not sure we should be doing this here as it implies the package knowing about the Jetpack plugin.
I'd consider removing the `j` param completely and make sure the Jetpack plugin adds it via a corresponding filter.
However before doing that we need to check how it will affect the backend tracking functionality on WPCOM

fgiannar commented on projects/packages/stats/tests/php/test-base.php:
```diff
@@ -22,7 +23,18 @@ abstract class StatsBaseTestCase extends BaseTestCase {
+	 *
+	 * @var string
+	 */
+	const DEFAULT_JETPACK__API_VERSION = '11.4';
```
See my comment above about removing this completely :) Ideally the package should have no knowledge about the Jetpack plugin

fgiannar reviewed (commented):
Many thanks for working on this, Juanma!

Left a couple of comments inline :)

darssen committed:
Renamed static function to stats_array_to_string to add meaning.

darssen committed:
Added docblock to filter

darssen committed:
Renamed constant to be more meaningful

darssen committed:
Merge branch 'trunk' into add/stats-tracking-pixel-to-pkg

brbrr commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,173 @@
+		$data = self::build_view_data();
+		// TODO Is this really needed here?
+		if ( class_exists( 'Jetpack_AMP_Support' ) && \Jetpack_AMP_Support::is_amp_request() ) {
+
```
redundant line? And there are a few more similar cases in this file.

darssen replied:
Do you mean the empty line?

brbrr replied:
yes :)

brbrr commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,173 @@
+		$blog_url = wp_parse_url( site_url() );
+		$srv      = $blog_url['host'];
+		$j        = sprintf( '%s:%s', Constants::get_constant( 'JETPACK__API_VERSION' ), Constants::get_constant( 'JETPACK__VERSION' ) );
+		if ( $wp_the_query->is_single || $wp_the_query->is_page || $wp_the_query->is_posts_page ) {
```
Should we test this if and try/finally block below?

darssen replied:
I guess we can test the **finally** section by checking that the `queried_object` and `queried_object_id` are not being changed. Should that be sufficient? 
The try functionality is tested here `test_build_view_data_with_post`

brbrr commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,173 @@
+	</script>
+END;
+		// phpcs:enable
+		print $stats_footer; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
Is there a reason why the rendering approach in `render_amp_footer` is different?

darssen replied:
Regarding if it makes sense to do the `get::footer `method, for me it makes total sense and was my initial approach. That way the testing is way easier. But then we discussed that it might be better to keep them for backward compatibility. I am not sure how to check if those methods are being used or not directly, aside from the stats module itself.

brbrr replied:
oh, right. in this case, we could leave the rendering methods as is but extract the composition logic into separate methods

darssen committed:
Removed unnecessary blank lines

darssen committed:
Merge branch 'add/stats-tracking-pixel-to-pkg' of github.com:Automattic/jetpack into add/stats-tracking-pixel-to-pkg

darssen committed:
Renaming since the function was renamed in a prior commit.

darssen committed:
Extracted the composing part of render functions to be able to test them easily.
AMP logic has been encapsulated as well.
Changed access modifiers for the internal methods to private. Left the render methods public for backward compatibility.

darssen committed:
Renamed test methods

darssen commented on projects/packages/stats/src/class-tracking-pixel.php:
```diff
@@ -0,0 +1,198 @@
+	 */
+	public static function add_to_footer() {
+		$data   = self::build_view_data();
+		$footer = get_footer_to_add( $data );
```
Extracted functionality as suggested so that it is easier to test.

darssen reviewed (commented):
Refactored code to be more testable.
- Created the `get_footer_to_add` method that has the composing logic for the footer and also encapsulates the AMP. This method returns the information in a string instead of already embedding it so the result can be tested properly.
- Changed access modifiers of `get_footer`, `get_amp_footer` and `stats_array_to_string` to **private**. 
- Mantained `render_footer` and `render_amp_footer` as public for backward compatibility.
- Added **Mockery** to the tests dependencies to be able to test the `\Jetpack_AMP_Support::is_amp_request()` check.

darssen committed:
Removed j param since the package shouldn't need to know about the Jetpack plugin.

darssen committed:
Added private function to check for is_amp_request instead of relying on Jetpack_AMP_Support.
Removed Mockery code and dependency since it is not needed.

darssen committed:
Merge branch 'trunk' into add/stats-tracking-pixel-to-pkg

fgiannar reviewed (approved):
LGTM!
